File: Question/53_Maximum_Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.empty()) return 0;
        vector<int> dp(nums.size(), 0);
        dp[0] = nums[0];
        for(int i = 1; i < nums.size(); ++i) {
            dp[i] = dp[i - 1] <= 0 ? nums[i] : nums[i] + dp[i - 1];
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: Question_Java/53_Maximum_Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int[] dp = new int[nums.length]; // dp[i] means max value uses until ith num
        dp[0] = nums[0];
        int ans = dp[0];
        for (int i = 1; i < nums.length; ++i) {
            dp[i] = dp[i - 1] <= 0 ? nums[i] : nums[i] + dp[i - 1];
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }
}